.github/workflows/flutter-deploy.yml

```diff
@@ -0,0 +1,120 @@
+        uses: actions/cache@v2
+        id: avd-cache
+        with:
+          path: |
+            ~/.android/avd/*
+            ~/.android/adb*
+          key: avd-31
+
+      - name: Ceate AVD and generate snapshot for caching
+        if: steps.avd-cache.outputs.cache-hit != 'true'
+        uses: reactivecircus/android-emulator-runner@v2
+        with:
+          api-level: 31
+          target: google_apis
+          arch: x86_64
+          force-avd-creation: false
+          emulator-options: -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
+          disable-animations: false
+          script: echo "Generated AVD snapshot for caching."
+
+      - name: Setup Flutter 2.10.1
+        uses: subosito/flutter-action@v2
+        with:
+          flutter-version: "2.10.1"
+          channel: "stable"
+          cache: true
+          cache-key: flutter-2.10.1
+          cache-path: ${{ runner.tool_cache }}/flutter
+
+      - name: Create .env file
+        run: |
+          echo "NEWSURL=https://www.freecodecamp.org/news/ghost/api/v3/content/" > .env
+          echo "NEWSKEY=$GHOST_NEWSKEY" >> .env
+          echo "ALGOLIAAPPID=$ALGOLIA_APP_ID" >> .env
+          echo "ALGOLIAKEY=$ALGOLIA_KEY" >> .env
+        env:
+          GHOST_NEWSKEY: ${{ secrets.GHOST_NEWSKEY }}
+          ALGOLIA_APP_ID: ${{ secrets.ALGOLIA_APP_ID }}
+          ALGOLIA_KEY: ${{ secrets.ALGOLIA_KEY }}
+
+      - name: Install packages
+        run: flutter pub get
+
+      - name: Analyze code
+        run: flutter analyze
+
+      - name: Run tests in emulator
+        uses: reactivecircus/android-emulator-runner@v2
+        with:
+          api-level: 31
+          target: google_apis
+          arch: x86_64
+          force-avd-creation: false
+          emulator-options: -no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
+          disable-animations: true
+          script: cd mobile-app && if adb shell pm list packages | grep -q org.freecodecamp; then adb uninstall org.freecodecamp; fi && flutter drive --driver=test_driver/integration_test.dart --target=integration_test/news_test.dart
+
+      - name: Upload screenshots
+        uses: actions/upload-artifact@v2
+        if: always()
+        with:
+          name: screenshots
+          path: mobile-app/screenshots/
+          retention-days: 7
+
+      - name: Setup Ruby for Fastlane
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: "2.6"
+          bundler-cache: true
+          working-directory: ./mobile-app/android
+
+      - name: Configure keystore
+        working-directory: ./mobile-app/android
+        run: |
+          echo "$KEYSTORE" | base64 --decode > app/keystore.jks
+          echo "$GPLAY_STORE_CONFIG" | base64 --decode > play-store-credentials.json
+          echo "storeFile=keystore.jks" > key.properties
+          echo "storePassword=$KEYSTORE_STORE_PASSWORD" >> key.properties
```

Comment:
This feels like something that shouldn't be in CI.  Is it possible to use a token instead?  Ideally that token would just have permission to upload to the play store.

Reply:
https://developers.google.com/android-publisher/authorization

From what I can tell, there aren't granular permissions for an API token and it would essentially be the same concern.

Reply:
Without the keystore and its details the app bundle cannot be signed for deployment. Also the credentials are for the service account which is needed for uploading the signed app bundle.

Reply:
This is fine. We are using a service (non-human) account - which has limited access to features and can be rotated in times of need.

Reply:
Great, thanks for confirming.  In that case, this looks look.  I won't pretend I understand every detail, but nothing else looked concerning.